src/jit/flowgraph.cpp

```diff
@@ -1621,15 +1694,152 @@ void Compiler::fgReplacePred(BasicBlock* block, BasicBlock* oldPred, BasicBlock*
     noway_assert(newPred != nullptr);
     assert(!fgCheapPredsValid);
 
-    flowList* pred;
+    // Make sure the list is sorted before manipulating.
+    bool sorted = fgExpensiveSortedPredCheck(block);
 
-    for (pred = block->bbPreds; pred != nullptr; pred = pred->flNext)
+#if DEBUG
+    // Also make sure that our refCount is correct before modifying.
+    // If it is not, keep track of that so that we know replacePreds is
+    // not the problem.
```

Comment:
I presume there are cases currently where the ref count is not correct?

There should be a TODO of some kind here to either add code like `assert(fgRefCountMatchesPredRefCount(block))`, or similar. I would hope either this whole DEBUG clause goes away, or gets replaced by a function call / assert.